clients/filesystem-hadoop/src/main/java/com/datastrato/gravitino/filesystem/hadoop/GravitinoVirtualFileSystem.java

```diff
@@ -0,0 +1,385 @@
+/*
+ * Copyright 2024 Datastrato Pvt Ltd.
+ * This software is licensed under the Apache License version 2.
+ */
+package com.datastrato.gravitino.filesystem.hadoop;
+
+import com.datastrato.gravitino.Catalog;
+import com.datastrato.gravitino.NameIdentifier;
+import com.datastrato.gravitino.client.GravitinoClient;
+import com.datastrato.gravitino.client.GravitinoMetaLake;
+import com.datastrato.gravitino.file.Fileset;
+import com.datastrato.gravitino.shaded.com.google.common.annotations.VisibleForTesting;
+import com.datastrato.gravitino.shaded.com.google.common.base.Preconditions;
+import com.datastrato.gravitino.shaded.com.google.common.cache.Cache;
+import com.datastrato.gravitino.shaded.com.google.common.cache.CacheBuilder;
+import com.datastrato.gravitino.shaded.com.google.common.cache.RemovalListener;
+import com.datastrato.gravitino.shaded.org.apache.commons.lang3.StringUtils;
+import java.io.IOException;
+import java.net.URI;
+import java.util.Arrays;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.FSDataInputStream;
+import org.apache.hadoop.fs.FSDataOutputStream;
+import org.apache.hadoop.fs.FileStatus;
+import org.apache.hadoop.fs.FileSystem;
+import org.apache.hadoop.fs.InvalidPathException;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.fs.permission.FsPermission;
+import org.apache.hadoop.util.Progressable;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * {@link GravitinoVirtualFileSystem} is a virtual file system which users can access `fileset` and
+ * other resources. It obtains the actual storage location corresponding to the resource from the
+ * Gravitino server, and creates an independent file system for it to act as an agent for users to
+ * access the underlying storage.
+ */
+public class GravitinoVirtualFileSystem extends FileSystem {
+  private static final Logger Logger = LoggerFactory.getLogger(GravitinoVirtualFileSystem.class);
+  private static final int DEFAULT_CACHE_CAPACITY = 20;
+  private static final int CACHE_EXPIRE_AFTER_ACCESS_MINUTES = 5;
```

Comment:
Can these two variations be made tunable?

Reply:
Yeah, i think it can be used as a customizable parameter.

Reply:
Done